File: packages/php/src/Resources/Tasks.php
<?php

declare(strict_types=1);

namespace MissionInbox\Resources;

use MissionInbox\Http\Client;

/**
 * The `tasks` resource. Access via `$mi->tasks`.
 *
 * Bulk operations dispatch background tasks and return `['taskId' => ...]`.
 * Poll to completion via {@see Tasks::waitFor()} or call {@see Tasks::get()}
 * yourself.
 */
final class Tasks
{
    private const TERMINAL_STATUSES = ['COMPLETED', 'FAILED', 'CANCELLED'];

    public function __construct(private readonly Client $http) {}

    /**
     * List tasks.
     *
     * @param array{type?: string, status?: string, page?: int, limit?: int} $params
     */
    public function list(array $params = []): array
    {
        return $this->http->request('GET', '/api/tasks', query: $params);
    }

    /** Retrieve a task by id. */
    public function get(string $id): array
    {
        return $this->http->request('GET', '/api/tasks/' . rawurlencode($id));
    }

    /** Cancel a pending or in-progress task. */
    public function cancel(string $id): array
    {
        return $this->http->request('DELETE', '/api/tasks/' . rawurlencode($id) . '/cancel');
    }

    /** Retrieve a task's execution log. `$since` is the task-output id from a previous poll. */
    public function getOutputs(string $id, ?string $since = null): array
    {
        $query = $since !== null ? ['since' => $since] : null;
        return $this->http->request(
            'GET',
            '/api/tasks/' . rawurlencode($id) . '/outputs',
            query: $query,
        );
    }

    /** Account-wide task counts by status. */
    public function getStatsSummary(): array
    {
        return $this->http->request('GET', '/api/tasks/stats/summary');
    }

    /**
     * Poll a task until it reaches a terminal state, then return the final task.
     * Throws {@see \RuntimeException} if the deadline passes first.
     *
     * @param array{
     *   pollInterval?: int,
     *   timeout?: int,
     *   onProgress?: callable(array<string, mixed>): void
     * } $options pollInterval + timeout are in milliseconds
     */
    public function waitFor(string $id, array $options = []): array
    {
        $pollInterval = $options['pollInterval'] ?? 2000;
        $timeout = $options['timeout'] ?? 300_000;
        $onProgress = $options['onProgress'] ?? null;
        $deadline = (int) (microtime(true) * 1000) + $timeout;

        while (true) {
            $task = $this->get($id);
            if ($onProgress !== null) {
                $onProgress($task);
            }
            $status = $task['status'] ?? null;
            if (is_string($status) && in_array($status, self::TERMINAL_STATUSES, true)) {
                return $task;
            }
            $now = (int) (microtime(true) * 1000);
            if ($now + $pollInterval > $deadline) {
                throw new \RuntimeException(sprintf(
                    'Timed out after %dms waiting for task %s (last status: %s)',
                    $timeout,
                    $id,
                    $status ?? 'unknown',
                ));
            }
            usleep($pollInterval * 1000);
        }
    }
}

File: packages/java/src/main/java/com/missioninbox/resources/Tasks.java
package com.missioninbox.resources;

import com.missioninbox.http.HttpClient;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Consumer;

/**
 * The {@code tasks} resource. Access via {@code mi.tasks}.
 *
 * <p>Bulk operations dispatch background tasks and return a map with a
 * {@code taskId}. Poll to completion via {@link #waitFor(String)} or call
 * {@link #get(String)} yourself.
 */
public final class Tasks {

    /** Terminal statuses reached by {@link #waitFor}. */
    public static final List<String> TERMINAL_STATUSES =
            Collections.unmodifiableList(Arrays.asList("COMPLETED", "FAILED", "CANCELLED"));

    private static final Set<String> TERMINAL_SET = new HashSet<>(TERMINAL_STATUSES);

    private final HttpClient http;

    public Tasks(HttpClient http) {
        this.http = http;
    }

    /**
     * List tasks. Recognised query keys: {@code type}, {@code status}, {@code page}, {@code limit}.
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> list(Map<String, Object> params) {
        return (Map<String, Object>) http.request("GET", "/api/tasks", params, null);
    }

    public Map<String, Object> list() {
        return list(null);
    }

    /** Retrieve a task by id. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> get(String id) {
        return (Map<String, Object>) http.request("GET", "/api/tasks/" + encode(id), null, null);
    }

    /** Cancel a pending or in-progress task. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> cancel(String id) {
        return (Map<String, Object>) http.request(
                "DELETE", "/api/tasks/" + encode(id) + "/cancel", null, null);
    }

    /** Retrieve a task's execution log. {@code since} is a task-output id from a previous poll. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> getOutputs(String id, String since) {
        Map<String, Object> query = since == null
                ? null
                : Collections.singletonMap("since", since);
        return (Map<String, Object>) http.request(
                "GET", "/api/tasks/" + encode(id) + "/outputs", query, null);
    }

    public Map<String, Object> getOutputs(String id) {
        return getOutputs(id, null);
    }

    /** Account-wide task counts by status. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> getStatsSummary() {
        return (Map<String, Object>) http.request("GET", "/api/tasks/stats/summary", null, null);
    }

    /**
     * Poll a task until it reaches a terminal state.
     *
     * @param id Task id.
     * @param pollIntervalMs Milliseconds between polls.
     * @param timeoutMs Total deadline in milliseconds.
     * @param onProgress Optional consumer called after each poll.
     * @throws RuntimeException with a "Timed out" message if the deadline passes.
     */
    public Map<String, Object> waitFor(
            String id,
            long pollIntervalMs,
            long timeoutMs,
            Consumer<Map<String, Object>> onProgress) {
        long deadline = System.currentTimeMillis() + timeoutMs;
        while (true) {
            Map<String, Object> task = get(id);
            if (onProgress != null) onProgress.accept(task);
            Object status = task.get("status");
            if (status instanceof String && TERMINAL_SET.contains(status)) {
                return task;
            }
            if (System.currentTimeMillis() + pollIntervalMs > deadline) {
                throw new RuntimeException(String.format(
                        "Timed out after %dms waiting for task %s (last status: %s)",
                        timeoutMs, id, status));
            }
            try {
                Thread.sleep(pollIntervalMs);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException("Interrupted while waiting for task " + id, e);
            }
        }
    }

    public Map<String, Object> waitFor(String id) {
        return waitFor(id, 2_000, 300_000, null);
    }

    private static String encode(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }
}
